### PR TITLE
Ship versioned libc headers on macOS

### LINKED ISSUE #10200: aarch64-macos-gnu, any-macos-any headers need to be updated for macOS 12.0 Monterey
### Zig Version

0.9.0-dev.1696+ff9a44a9f

### Steps to Reproduce

I believe we need to regenerate the macOS libc headers to include changes from macOS 12.0 Monterey.

---

(1) libc headers for aarch64-macos-gnu [report only macOS 11.1 being available](https://github.com/ziglang/zig/blob/b644d493658d8fe4eddeeb3ec6d13593f89739bb/lib/libc/include/aarch64-macos-gnu/AvailabilityVersions.h#L61) (12.0 Monterey is the newest), which can cause issues when trying to compile some code. ~~I believe I'm facing this in https://github.com/hexops/mach/issues/84~~

(2) [`sys/cdefs.h`](https://github.com/ziglang/zig/blob/master/lib/libc/include/any-macos-any/sys/cdefs.h) currently does not define these as empty, which the latest header file from the macOS 12.0 SDK does:

```
#define __kernel_ptr_semantics
#define __kernel_data_semantics
#define __kernel_dual_semantics
``` 

This makes it impossible to supply macOS 12.0 frameworks and compile their headers, as e.g. definitions in `IOKit/IOTypes.h` become syntactically invalid:

```
typedef mach_vm_address_t       IOVirtualAddress __kernel_ptr_semantics;
```

### Expected Behavior

The headers reflect macOS 12

### Actual Behavior

They do not reflect macOS 12

### PR DESCRIPTION
**Pull request by kubkon**
Fixes #10200 

Previously, we would ship a singular set of libc headers per macOS CPU architecture. This PR rectifies it by now shipping a separate set of headers per macOS version AND CPU architecture. The headers are deduplicated among themselves following a 3 layer approach such that:

* layer 1: `x86_64-macos.10` `x86_64-macos.11` `x86_64-macos.12` `aarch64-macos.11` `aarch64-macos.12`
* layer 2: `any-macos.10` `any-macos.11` `any-macos.12`
* layer 3: `any-macos-any`

where each subsequent layer contains common headers from the previous layer. More on this can be found in the discussion in https://github.com/ziglang/fetch-them-macos-headers/pull/15.

Changes required to make this work include:
- [x] treating macOS target separately in `Compilation.detectLibCIncludeDirs` (https://github.com/ziglang/zig/commit/dab641442fa8938b8414507346a244d0a9796ff7)
- [x] temporarily tweaking `std.Target.Os.defaultVersionRange` to also accept `std.Target.Cpu.Arch` enum as there is currently a mismatch between the minimum version of macOS supported by `x86_64` and `aarch64` (the former is 10.15, while the latter is 11.6) (https://github.com/ziglang/zig/commit/baf4609045c770f41e566d129b88e3fdb309e5c5)
- [x] do not rely on headers shipped with Zig if we are building natively AND we have detected an SDK using `xcrun --sdk macosx --show-sdk-path` - turns out that it is possible to have SDK 11.1 on macOS 10.15, or SDK 12.0 on macOS 11.6. I'm guessing that Apple offers devs on the older OSes to target newer / latest OS too.
- [x] version `libSystem.B.tbd` shipped with Zig to match the versions that Zig currently supports

### CODE REVIEW

**Comment by kubkon** [comment]
cc @slimsag

**Comment by andrewrk** [comment]
Looks good so far! There's one more place you'll need to edit in Compilation.zig for the generated builtin.zig source.

**Comment by kubkon** [comment]
> Looks good so far! There's one more place you'll need to edit in Compilation.zig for the generated builtin.zig source.

Hmm, are you sure anything needs updating in `Compilation.zig`? I don't see any use of `defaultVersionRange` in the generated `builtin.zig` source there. It seems to be using a direct constructor with already inferred version ranges for the OS.

**Comment by kubkon** [comment]
Ready for review! I've fixed one of the oversights I made when adding better iOS, iPadOS, etc. support to Zig. Namely, since all of these represent cross-compilation targets from macOS, if you decide to cross-compile to any such target, you will need to manually specify relevant sysroot and syslib / framework paths on the command line and in your `build.zig` file. This behaviour is now consistent with any other cross-compilation target in Zig I believe. The auto-inferred SDK paths are only used if we're compiling something natively (from host to host). So, to build for iPhone simulator on the M1 for example, you will need to spell out something like:

```
$ zig build-exe test.zig -target aarch64-ios-simulator --sysroot $(xcrun --sdk iphonesimulator --show-sdk-path) -L/usr/lib -F/System/Library/Frameworks
```

But, if you're building the same `test.zig` natively, this suffices:

```
$ zig build-exe test.zig
```

**Comment by kubkon** [comment]
Finally, you are now free to target any of the supported macOS versions for cross-compilation using version ranges. For instance, in order to target Catalina from Monterey, all you need to do is:

```
$ zig cc hello.c -target x86_64-macos.10
```

Similarly, if we want to target M1 Monterey from a Catalina-based Intel MBP, we can do:

```
$ zig cc hello.c -target aarch64-macos.12
```

Tl;dr versioning is currently based on major macOS revisions - 10 for Catalina, 11 for BigSur, 12 for Monterey. If this is confusing, or if Apple decides to use minor versioning again between different OS cuts (10.14 vs 10.15), we can add that in too.

File: src/Compilation.zig [hunk 1]
```diff
@@ -962,15 +962,17 @@ pub fn create(gpa: *Allocator, options: InitOptions) !*Compilation {
...
+            if (builtin.os.tag != .macos or !options.target.isDarwin()) break :blk false;
+            break :blk options.frameworks.len > 0 or options.framework_dirs.len > 0;
         };
 
         const sysroot = blk: {
             if (options.sysroot) |sysroot| {
                 break :blk sysroot;
-            } else if (darwin_can_use_system_sdk) {
+            } else if (darwin_use_system_sdk) {
                 break :blk try std.zig.system.darwin.getSDKPath(arena, options.target);
```
Comment (slimsag) [inline src/Compilation.zig:976]: I think that this change _might_ fix an issue I've been meaning to file for a while now, https://github.com/ziglang/zig/issues/10217, specifically when specifying a cross-compilation target. WDYT?

**Review by slimsag — APPROVED** [review]
Looks quite good to me!

File: src/Compilation.zig (hunk 1 above)
Comment (kubkon) [reply src/Compilation.zig:976]: Yeah, I hope that's the case too actually. Would you mind checking after this gets merged, or building zig directly from this branch?

Comment (slimsag) [reply src/Compilation.zig:976]: Will do!

**Comment by andrewrk** [comment]
CI failure:

```
error(link): undefined reference to symbol '___ulock_wait2'
error(link):   first referenced in '/workspace/zig-cache/o/f13b7e43e90bbe6475a7f932d7504a84/test.o'
thread 51767 panic: attempt to unwrap error: UndefinedSymbolReference
/workspace/src/main.zig:216:9: 0x30c4488 in main.mainArgs (zig1)
        return buildOutputType(gpa, arena, args, .zig_test);
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
        return error.OutOfMemory;
        ^
/workspace/src/link/MachO.zig:890:13: 0x32a8643 in link.MachO.flushModule (zig1)
            return error.UndefinedSymbolReference;
            ^
/workspace/src/link/MachO.zig:401:5: 0x328b287 in link.MachO.flush (zig1)
    try self.flushModule(comp);
    ^
/workspace/src/link.zig:541:23: 0x3253083 in link.File.flush (zig1)
            .macho => return @fieldParentPtr(MachO, "base", base).flush(comp),
                      ^
/workspace/src/Compilation.zig:1877:5: 0x324b651 in Compilation.update (zig1)
    try self.bin_file.flush(self);
    ^
/workspace/src/main.zig:2689:5: 0x31d90ed in main.updateModule (zig1)
    try comp.update();
    ^
/workspace/src/main.zig:2372:21: 0x30f147a in main.buildOutputType (zig1)
        else => |e| return e,
                    ^
???:?:?: 0x30c5c22 in ??? (???)
/workspace/src/stage1.zig:48:43: 0x30c3d12 in main (zig1)
        stage2.mainArgs(gpa, arena, args) catch unreachable;
                                          ^
???:?:?: 0x8379fe8 in ??? (???)
The following command terminated unexpectedly:
/workspace/_debug/staging/bin/zig test /workspace/lib/std/std.zig --test-name-prefix std-x86_64-macos-gnu-Debug--multi  --cache-dir /workspace/zig-cache --global-cache-dir /root/.cache/zig --name test -target x86_64-macos-gnu -mcpu x86_64 -I /workspace/test --zig-lib-dir /workspace/lib 
```

**Comment by kubkon** [comment]
> CI failure:
> 
> ```
> error(link): undefined reference to symbol '___ulock_wait2'
> error(link):   first referenced in '/workspace/zig-cache/o/f13b7e43e90bbe6475a7f932d7504a84/test.o'
> thread 51767 panic: attempt to unwrap error: UndefinedSymbolReference
> /workspace/src/main.zig:216:9: 0x30c4488 in main.mainArgs (zig1)
>         return buildOutputType(gpa, arena, args, .zig_test);
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/lib/std/heap.zig:150:9: 0x30be120 in std.heap.CAllocator.resize (zig1)
>         return error.OutOfMemory;
>         ^
> /workspace/src/link/MachO.zig:890:13: 0x32a8643 in link.MachO.flushModule (zig1)
>             return error.UndefinedSymbolReference;
>             ^
> /workspace/src/link/MachO.zig:401:5: 0x328b287 in link.MachO.flush (zig1)
>     try self.flushModule(comp);
>     ^
> /workspace/src/link.zig:541:23: 0x3253083 in link.File.flush (zig1)
>             .macho => return @fieldParentPtr(MachO, "base", base).flush(comp),
>                       ^
> /workspace/src/Compilation.zig:1877:5: 0x324b651 in Compilation.update (zig1)
>     try self.bin_file.flush(self);
>     ^
> /workspace/src/main.zig:2689:5: 0x31d90ed in main.updateModule (zig1)
>     try comp.update();
>     ^
> /workspace/src/main.zig:2372:21: 0x30f147a in main.buildOutputType (zig1)
>         else => |e| return e,
>                     ^
> ???:?:?: 0x30c5c22 in ??? (???)
> /workspace/src/stage1.zig:48:43: 0x30c3d12 in main (zig1)
>         stage2.mainArgs(gpa, arena, args) catch unreachable;
>                                           ^
> ???:?:?: 0x8379fe8 in ??? (???)
> The following command terminated unexpectedly:
> /workspace/_debug/staging/bin/zig test /workspace/lib/std/std.zig --test-name-prefix std-x86_64-macos-gnu-Debug--multi  --cache-dir /workspace/zig-cache --global-cache-dir /root/.cache/zig --name test -target x86_64-macos-gnu -mcpu x86_64 -I /workspace/test --zig-lib-dir /workspace/lib 
> ```

A bug in libstd, fixed in a041401.

**Review by andrewrk — APPROVED** [review]
I think this is a really big improvement for targeting macOS!

For those who aren't paying super close attention (and for myself when I type up release notes in a couple weeks):

### Master branch, before this PR

 * macOS C headers & .tbd files total installation size: 6.1 MiB
   - tarball size: 534 KB
 * supported triples:
   - x86_64-macos.11 (Big Sur)
   - aarch64-macos.11 (Big Sur)

### After this PR is merged

 * macOS C headers & .tbd files total installation size: 13.1 MiB
   - tarball size: 581 KB
 * supported triples:
   - x86_64-macos.10 (Catalina)
   - x86_64-macos.11 (Big Sur)
   - x86_64-macos.12 (Monterey)
   - aarch64-macos.11 (Big Sur)
   - aarch64-macos.12 (Monterey)
